### ISSUE TITLE
Create a generator with a producer function without macros with types different to Unit

### ISSUE THREAD

**Issue by PatatasDelPapa** [whatisaphone/genawaiter#29]
TLDR: I want to create the same generators as the macros using the low level api, specifically with a named function, but I can't

Note: All of this is using the `sync` module

I've been poking around this library for a bit and found out that creating a genawaiter::sync::Gen<u32, u32, impl Future> is easy with macros as it's only a little modification from the example of the [docs](https://docs.rs/genawaiter/0.99.1/genawaiter/sync/index.html#passing-resume-arguments)
```rust
fn main() {
    let mut example_generator = gen!({
        let num = yield_!(42_u32);
        let num = yield_!(42);
    });
    
    example_generator.resume_with(0_u32);
    example_generator.resume_with(1);
    example_generator.resume_with(2);
}
```
Checking the type is easy. By adding a resume() this will not compile and the error will tell the type of the generator
```rust
example_generator.resume_with(0_u32);
example_generator.resume_with(1);
example_generator.resume_with(2);
example_generator.resume();
```
```
error[E0599]: no method named `resume` found for struct `genawaiter::sync::Gen<u32, u32, impl Future>` in the current scope
  --> src\main.rs:13:23
   |
13 |     example_generator.resume();
   |                       ^^^^^^ method not found in `genawaiter::sync::Gen<u32, u32, impl Future>`
```
Then I played with the low level api as I like to see how would it be to create the same generator without macros. I follow the [docs](https://docs.rs/genawaiter/0.99.1/genawaiter/sync/index.html#using-the-low-level-api-with-function-arguments) for the sync low level api and so I define my async producer function and use it like so.

```rust
async fn producer(n: u32, co: Co<u32>) {
    let num = co.yield_(n).await;
    let num = co.yield_(n).await;
}

fn main() {
    let mut example_generator = Gen::new(|co| producer(42, co));
    example_generator.resume_with(0_u32);
    example_generator.resume_with(1);
    example_generator.resume_with(2);
}
```
But I got errors
```
error[E0308]: mismatched types
  --> src\main.rs:13:35
   |
13 |     example_generator.resume_with(0_u32);
   |                                   ^^^^^ expected `()`, found `u32`

error[E0308]: mismatched types
  --> src\main.rs:14:35
   |
14 |     example_generator.resume_with(1);
   |                                   ^ expected `()`, found integer

error[E0308]: mismatched types
  --> src\main.rs:15:35
   |
15 |     example_generator.resume_with(2);
   |                                   ^ expected `()`, found integer

error: aborting due to 3 previous errors
```
Looking around I found this definition
```rust
type Co<Y, R = ()> = Co<Airlock<Y, R>>;
```
To do the same as the macros my Co needs to be `Co<u32, u32>` but instead is `Co<u32, ()>` but I can't change it and I didn't found another public and more flexible Co. If I want to use `return_with` with a type different to () I simply can't use named functions?

PD: This also works and don't use macros but I can't use a named function
```rust
fn main() {
    let mut example_generator = Gen::new(|co| async move {
        let num = co.yield_(42_u32).await;
        let num = co.yield_(42).await;
    });
    example_generator.resume_with(0_u32);
    example_generator.resume_with(1);
    example_generator.resume_with(2);
}
```

**Comment by DevinR528** [comment]
The `R = ()` is the default type, you can specify your own type there and get the named function example working.
```rust
async fn producer(n: u32, co: Co<u32, u32>) {
    println!("first {}", n);
    let num = co.yield_(n).await;
    println!("sec {}", num);

    let num = co.yield_(n).await;
    println!("sec {}", num);

    let num = co.yield_(n).await;
    println!("sec {}", num);
}

let mut example_generator = Gen::new(|co| producer(42, co));
example_generator.resume_with(0_u32); // first call is lost (explained in sync mod docs)
example_generator.resume_with(0_u32);
example_generator.resume_with(1);
example_generator.resume_with(2);
```

**Comment by PatatasDelPapa** [comment]
> The R = () is the default type, you can specify your own type there and get the named function example working.
 
I don't know why I didn't try that and now I feel bad. Either way thanks for the clarification

**Comment by DevinR528** [comment]
No worries!
